File: adapter/src/main/java/com/github/lany192/adapter/OneToManyBuilder.java
package com.github.lany192.adapter;

import static com.github.lany192.adapter.Preconditions.checkNotNull;

import androidx.annotation.CheckResult;
import androidx.annotation.NonNull;


class OneToManyBuilder<T> implements OneToManyFlow<T>, OneToManyEndpoint<T> {
    @NonNull
    private final MultiTypeAdapter adapter;
    @NonNull
    private final Class<? extends T> clazz;
    private ItemViewBinder<T, ?>[] binders;

    OneToManyBuilder(@NonNull MultiTypeAdapter adapter, @NonNull Class<? extends T> clazz) {
        this.clazz = clazz;
        this.adapter = adapter;
    }

    @Override
    @CheckResult
    @SafeVarargs
    public final @NonNull
    OneToManyEndpoint<T> to(@NonNull ItemViewBinder<T, ?>... binders) {
        checkNotNull(binders);
        this.binders = binders;
        return this;
    }


    @Override
    public void withLinker(@NonNull Linker<T> linker) {
        checkNotNull(linker);
        doRegister(linker);
    }


    @Override
    public void withClassLinker(@NonNull ClassLinker<T> classLinker) {
        checkNotNull(classLinker);
        doRegister(ClassLinkerWrapper.wrap(classLinker, binders));
    }


    private void doRegister(@NonNull Linker<T> linker) {
        for (ItemViewBinder<T, ?> binder : binders) {
            adapter.register(clazz, binder, linker);
        }
    }
}

File: library/src/main/java/com/github/lany192/box/multitype/OneToManyBuilder.kt
package com.github.lany192.box.multitype

import androidx.annotation.CheckResult


internal class OneToManyBuilder<T>(
    private val adapter: MultiTypeAdapter,
    private val clazz: Class<T>
) : OneToManyFlow<T>, OneToManyEndpoint<T> {

    private var delegates: Array<ItemViewDelegate<T, *>>? = null

    @SafeVarargs
    @CheckResult(suggest = "#withLinker(Linker)")
    override fun to(vararg delegates: ItemViewDelegate<T, *>) = apply {
        @Suppress("UNCHECKED_CAST")
        this.delegates = delegates as Array<ItemViewDelegate<T, *>>
    }

    @SafeVarargs
    @CheckResult(suggest = "#withLinker(Linker)")
    override fun to(vararg binders: ItemViewBinder<T, *>) = apply {
        @Suppress("UNCHECKED_CAST")
        this.delegates = binders as Array<ItemViewDelegate<T, *>>
    }

    override fun withLinker(linker: Linker<T>) {
        doRegister(linker)
    }

    override fun withJavaClassLinker(javaClassLinker: JavaClassLinker<T>) {
        withLinker(ClassLinkerBridge.toLinker(javaClassLinker, delegates!!))
    }

    private fun doRegister(linker: Linker<T>) {
        for (delegate in delegates!!) {
            adapter.register(Type(clazz, delegate, linker))
        }
    }
}
